rust-src/aggregate_sig/src/aggregate_sig.rs

```diff
@@ -171,6 +171,52 @@ pub fn verify_aggregate_sig<P: Pairing>(
     P::pair(&signature.0, &P::G2::one_point()) == product
 }
 
+/// Verifies an aggregate signature on pairs `(messages m_i, set_i)` `for
+/// i=1..n` but where `set_i` denotes the set of public keys corresponding to
+/// the secret keys that signed m_i. This implements a combination of
+/// AggregateVerify from Section 3.1.1 and FastAggregateVerify from Section
+/// 3.3.4 of https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-05#section-3.1.1.
+///
+/// In particular, this function returns:
+///
+/// pairing(sig, g_2) == ∏ᵢ pairing(H(mᵢ), ∑ⱼ pkᵢⱼ)
+///
+/// Each message should be associated with at least one key. However, the result
+/// of having a message with no keys will be the same as if the message was not
+/// included.
+///
+/// For security, the holder of a key must be required to prove knowledge of the
+/// secret key.
+pub fn verify_aggregate_sig_hybrid<P: Pairing>(
+    m_pk_pairs: &[(&[u8], &[PublicKey<P>])],
+    signature: Signature<P>,
+) -> bool {
+    let product = m_pk_pairs
+        .par_iter()
+        .fold(<P::TargetField as Field>::one, |prod, (m, pks)| {
+            let sum_pk_i = if pks.len() < 150 {
+                pks.iter()
+                    .fold(P::G2::zero_point(), |s, x| s.plus_point(&x.0))
+            } else {
+                pks.par_iter()
+                    .fold(P::G2::zero_point, |s, x| s.plus_point(&x.0))
+                    .reduce(P::G2::zero_point, |s, x| s.plus_point(&x))
```

Comment:
Why do we here fold and then reduce? If that's not supposed to be obvious, maybe add a comment.

Reply:
It's because we use the rayon parallel itererator as opposed to the sequential `iter`.
Hence we `reduce` with the identity for the addition operator so rayon can do its thing.

Reply:
Thanks! From the documentation you linked, it looks like one can directly do the reduce. Are there performance implications on this or am I missing anything?

Reply:
You cannot directly reduce because the type you are iterating over is PUblicKey<P>, but the result is a group element.

You either first have to map to unwrap the public key or do what is done here.